pkg/common_controller/snapshot_controller.go

```diff
@@ -502,362 +645,270 @@ func (ctrl *csiSnapshotController) isVolumeBeingCreatedFromSnapshot(snapshot *cr
+		return nil
 	}
 
-	// Create VolumeSnapshotContent name
-	contentName := GetSnapshotContentNameForSnapshot(snapshot)
+	pvcClone := pvc.DeepCopy()
+	pvcClone.ObjectMeta.Finalizers = slice.RemoveString(pvcClone.ObjectMeta.Finalizers, utils.PVCFinalizer, nil)
 
-	// Resolve snapshotting secret credentials.
-	snapshotterSecretRef, err := getSecretReference(class.Parameters, contentName, snapshot)
+	_, err = ctrl.client.CoreV1().PersistentVolumeClaims(pvcClone.Namespace).Update(pvcClone)
 	if err != nil {
-		return nil, nil, "", nil, err
+		return newControllerUpdateError(pvcClone.Name, err.Error())
 	}
 
-	return class, volume, contentName, snapshotterSecretRef, nil
+	klog.V(5).Infof("Removed protection finalizer from persistent volume claim %s", pvc.Name)
+	return nil
 }
 
-func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
-	var err error
-	var creationTime time.Time
-	var size int64
-	var readyToUse = false
-	var driverName string
-	var snapshotID string
-
-	if snapshot.Spec.Source.VolumeSnapshotContentName != nil {
-		klog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: checking whether snapshot [%s] is pre-bound to content [%s]", snapshot.Name, content.Name)
-		readyToUse, creationTime, size, err = ctrl.handler.GetSnapshotStatus(content)
-		if err != nil {
-			klog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call get snapshot status to check whether snapshot is ready to use %q", err)
-			return nil, err
-		}
-		driverName, snapshotID = content.Spec.Driver, *content.Status.SnapshotHandle
-	} else {
-		class, volume, _, snapshotterSecretRef, err := ctrl.getCreateSnapshotInput(snapshot)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get input parameters to create snapshot %s: %q", snapshot.Name, err)
-		}
-		snapshotterCredentials, err := getCredentials(ctrl.client, snapshotterSecretRef)
-		if err != nil {
-			return nil, err
-		}
-		driverName, snapshotID, creationTime, size, readyToUse, err = ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
-		if err != nil {
-			klog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call create snapshot to check whether the snapshot is ready to use %q", err)
-			return nil, err
-		}
-	}
-	klog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: driver %s, snapshotId %s, creationTime %v, size %d, readyToUse %t", driverName, snapshotID, creationTime, size, readyToUse)
-
-	if creationTime.IsZero() {
-		creationTime = time.Now()
-	}
-	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, content.Name, readyToUse, creationTime, size)
+// isSnapshotSourceBeingUsed checks if a PVC is being used as a source to create a snapshot
+func (ctrl *csiSnapshotCommonController) isSnapshotSourceBeingUsed(snapshot *crdv1.VolumeSnapshot) bool {
```

Comment:
It might be more clear to let this function only takes in a PVC pointer instead of snapshot.
What this function will be doing is to check whether there is any Snapshot within the namespace which is currently using the PVC.

Reply:
but we also need more info from the snapshot itself in this routine.